contracts/core/SequencerProofVerifier.sol

```diff
@@ -0,0 +1,133 @@
+        external
+        view
+    {
+        _verifySequencerSignature(
+            l2Height, appHash, keccak256(RLPWriter.writeList(l1header.header)), l1header.stateRoot, signature
+        );
+    }
+
+    /**
+     * @dev Prove that a given state is not part of a proof
+     * @dev this method is mainly used for packet timeouts, which is currently not implemented
+     */
+    function verifyNonMembership(bytes32, bytes calldata, Ics23Proof calldata) external pure {
+        revert MethodNotImplemented();
+    }
+
+    /**
+     * @inheritdoc ISignatureVerifier
+     * @dev verifies a chain of ICS23 proofs
+     * Each computed subroot starting from index 0 must match the value of the next proof (hence chained proofs).
+     * The cosmos SDK and ics23 support chained proofs to switch between different proof specs.
+     * Custom proof specs are not supported here. Only Iavl and Tendermint or similar proof specs are supported.
+     */
+    function verifyMembership(bytes32 appHash, bytes calldata key, bytes calldata value, Ics23Proof calldata proofs)
+        external
+        pure
+    {
+        // first check that the provided proof indeed proves the keys and values.
+        if (keccak256(key) != keccak256(proofs.proof[0].key)) {
+            revert InvalidProofKey();
+        }
+        if (keccak256(value) != keccak256(proofs.proof[0].value)) revert InvalidProofValue();
+        // proofs are chained backwards. First proof in the list (proof[0]) corresponds to the packet proof, meaning
+        // that can be checked against the next subroot value (i.e. ibc root). Once the first proof is verified,
+        // we can check the second that corresponds to the ibc proof, that is checked against the app hash (app root)
+        if (bytes32(proofs.proof[1].value) != _verify(proofs.proof[0])) revert InvalidPacketProof();
+        if (appHash != _verify(proofs.proof[1])) revert InvalidIbcStateProof();
+    }
+
+    function _verifySequencerSignature(
+        uint256 L2BlockNumber,
+        bytes32 AppHash,
+        bytes32 L1BlockHash,
+        bytes32 L1StateRoot,
+        bytes calldata signature
+    ) internal view {
+        // TODO: check that erecover should indeed ensure that the signature signs over the hash
+        if (
+            ECDSA.recover(keccak256(bytes.concat(bytes32(L2BlockNumber), AppHash, L1BlockHash, L1StateRoot)), signature)
+                != sequencer
+        ) {
+            revert InvalidSequencerSignature();
+        }
+    }
+
+    /**
+     * @dev Verifies an ICS23 proof through the root hash based on the provided proof.
+     * @dev This code was adapted from the ICS23 membership verification found here:
+     * https://github.com/cosmos/ics23/blob/go/v0.10.0/go/ics23.go#L36
+     * @param proof The ICS23 proof to be verified.
+     * @return computed The computed root hash.
+     */
+    function _verify(OpIcs23Proof calldata proof) internal pure returns (bytes32 computed) {
```

Comment:
this looks a lot like the one from OptimisticProofVerifier.sol, can we put them in a lib instead of duplicating the code?

Reply:
same with the one below

Reply:
yes, I need to rebase the branch & update SequencerProofVerifier, but I implemented abstracting this away in [this commit](https://github.com/open-ibc/vibc-core-smart-contracts/pull/198/commits/1ae7826a03ad729d20a063cc59912698c8297648) on the stacked pr branch. 

Note: I chose implementing this functionality in an abstract contract instead of a library since that would cause an extra runtime `DELEGATECALL` gas cost for these methods, since solidity implements libraries as delegate calls. This doesn't happen with abstract contracts since it's compiled into the bytecode of the inheriting contracts

Reply:
should be abstracted away into `AppstateVerifier` class now